src/Firestore/FirestoreClient.php

```diff
@@ -0,0 +1,526 @@
+
+        return new DocumentReference(
+            $this->connection,
+            $this->valueMapper,
+            $this->collection($this->pathId($this->parentPath($name))),
+            $name
+        );
+    }
+
+    /**
+     * Get a list of documents by their path.
+     *
+     * The number of results generated will be equal to the number of documents
+     * requested, except in case of error.
+     *
+     * Note that this method will **always** return instances of
+     * {@see Google\Cloud\Firestore\DocumentSnapshot}, even if the documents
+     * requested do not exist. It is highly recommended that you check for
+     * existence before accessing document data.
+     *
+     * Example:
+     * ```
+     * $documents = $firestore->documents([
+     *     'users/john',
+     *     'users/dave'
+     * ]);
+     * ```
+     *
+     * ```
+     * // To check whether a given document exists, use `DocumentSnapshot::exists()`.
+     * $documents = $firestore->documents([
+     *     'users/deleted-user'
+     * ]);
+     *
+     * foreach ($documents as $document) {
+     *     if (!$document->exists()) {
+     *         echo $document->id() . ' Does Not Exist';
+     *     }
+     * }
+     * ```
+     *
+     * @codingStandardsIgnoreStart
+     * @see https://cloud.google.com/firestore/docs/reference/rpc/google.firestore.v1beta1#google.firestore.v1beta1.Firestore.BatchGetDocuments BatchGetDocuments
+     * @codingStandardsIgnoreEnd
+     *
+     * @param string[]|DocumentReference[] $paths Any combination of string paths or DocumentReference instances.
+     * @param array $options Configuration options.
+     * @return DocumentSnapshot[]
+     */
+    public function documents(array $paths, array $options = [])
+    {
+        array_walk($paths, function (&$path) {
+            if ($path instanceof DocumentReference) {
+                $path = $path->name();
+            }
+
+            $path = $this->isRelative($path)
+                ? $this->fullName($this->projectId, $this->database, $path)
+                : $path;
+        });
+
+        $documents = $this->connection->batchGetDocuments([
+            'database' => $this->databaseName($this->projectId, $this->database),
+            'documents' => $paths,
+        ] + $options);
+
+        $pathKeys = array_flip($paths);
+        $out = [];
+        foreach ($documents as $document) {
+            $exists = isset($document['found']);
+            $data = $exists
+                ? $document['found'] + ['readTime' => $document['readTime']]
+                : ['readTime' => $document['readTime']];
+
+            $name = $exists
+                ? $document['found']['name']
+                : $document['missing'];
+
+            $snap = $this->createSnapshotWithData(
+                $this->valueMapper,
+                $this->document($name),
+                $data,
+                $exists
+            );
+
+            $key = $pathKeys[$name];
+            $out[$key] = $snap;
```

Comment:
This was the original implementation in the Java client. Unfortunately, it does not work the user requests the same document multiple times. Instead, I ended up walking over the user input again to construct the result.

Reply:
I've updated the sorting mechanism to take that into account.